welkin/util.py

```diff
@@ -107,3 +110,90 @@ def clean_datetime(dt: datetime) -> str:
         .isoformat(timespec="milliseconds")
         .replace("+00:00", "Z")
     )
+
+
+def find_model_id(instance: Union[Collection, Resource], model_name: str) -> str:
+    """Recursively traverse the `_parent` chain searching for a model id.
+
+    Args:
+        instance (Union[Collection, Resource]): The instanceect instance to inspect.
+        model_name (str): The class name of the model to find.
+
+    Raises:
+        AttributeError: If recursion ends without finding the model id.
+
+    Returns:
+        str: The model id.
+    """
+    body_id_key = f"{to_camel_case(model_name)}Id"
+
+    if instance.__class__.__name__ == model_name:
+        return instance.id
+    elif hasattr(instance, body_id_key):
+        return getattr(instance, body_id_key)
+    elif instance._parent:
+        return find_model_id(instance._parent, model_name)
+
+    raise AttributeError(
+        f"Cannot find {model_name} id. Model._parent chain ends in {instance}"
+    )
+
+
+def model_id(*models: Tuple[str]) -> Callable:
+    """Insert values for `model_id` arguments if not provided.
+
+    Args:
+        *models (Tuple[str]): The model names to search for.
+
+    Raises:
+        TypeError: If no ID is found and no arguments are provided.
+    """
+
+    def decorator(f: Callable):
+        @wraps(f)
+        def wrapper(self, *args, **kwargs) -> Callable:
+            outer_exc = None
+            for model in models:
+                key = f"{to_snake_case(model)}_id"
+
+                if not args and key not in kwargs:
+                    try:
+                        kwargs[key] = find_model_id(self, model)
+                    except AttributeError as exc:
+                        outer_exc = exc
+
+            try:
+                return f(self, *args, **kwargs)
+            except TypeError as exc:
+                # Raise from the outer `AttributeError` so we don't lose context.
+                raise exc from outer_exc
```

Comment:
The exception pattern here probably looks a little strange, so let me explain.

At the end of the chain, we do want to raise the normal `TypeError` that Python would raise if the model ID argument isn't filled. However, we don't want to lose the context so implementers have some hints on how to fix the issue, since this mechanism is doing a bit of recursive introspection. The end result is a traceback chain that includes both the `TypeError` and the `AttributeError`(s) like so:

```python
>>> from welkin.models import Patient
>>> Patient().Encounter().EncounterDisposition().get()
Traceback (most recent call last):
  File ".../welkin-health/welkin/util.py", line 164, in wrapper
    raise e from outer_exc
  File ".../welkin-health/welkin/util.py", line 161, in wrapper
    kwargs[key] = find_model_id(self, model)
  File ".../welkin-health/welkin/util.py", line 137, in find_model_id
    raise AttributeError(
AttributeError: Cannot find Patient id. Model._parent chain ends in EncounterDisposition

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File ".../welkin-health/welkin/util.py", line 164, in wrapper
    raise e from outer_exc
  File ".../welkin-health/welkin/util.py", line 161, in wrapper
    kwargs[key] = find_model_id(self, model)
  File ".../welkin-health/welkin/util.py", line 137, in find_model_id
    raise AttributeError(
AttributeError: Cannot find Encounter id. Model._parent chain ends in EncounterDisposition

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File ".../welkin-health/welkin/util.py", line 172, in wrapper
    raise exc from outer_exc
  File ".../welkin-health/welkin/util.py", line 169, in wrapper
    return f(self, *args, **kwargs)
TypeError: EncounterDisposition.get() missing 2 required positional arguments: 'patient_id' and 'encounter_id'
```